### ISSUE TITLE
Add generic overload for `IFakeOptionsBuilder<T>.Implements`

### ISSUE THREAD

**Issue by thomaslevesque** [FakeItEasy/FakeItEasy#470]
Currently, to implement an additional interface on a fake, we have to write this:

    A.Fake<IFoo>(o => o.Implements(typeof(IBar))

It would be more convenient to write it like this:

    A.Fake<IFoo>(o => o.Implements<IBar>())

It's a minor change, but it contributes to make the API more fluent IMO.

The only drawback is that this overload has to be added to the `IFakeOptionsBuilder<T>` interface, it can't be an extension method. The reason is that the extension method would need to have 2 type parameters, and since it's not possible to specify only one and let the compiler infer the other, we would have to specify both explicitly.

**Comment by thomaslevesque** [comment]
See PR #471

**Comment by blairconrad** [comment]
Thanks for suggesting this, @thomaslevesque.

Unfortunately, I'm going to be my usual self and question the value. 

I agreed that the syntax is cleaner (a little), but

1. it enables no new use cases, and 
2. would either have to replace the existing mechanism or provide a double for it

As I've recently been taking a fairly strong stance against introducing a second way of performing an operation when the new one lacks a sizable advantage over the original, I don't think I could advocate adding this mechanism as a second option for specifying `Implements`.

And as for replacing the original mechanism, while we've been on a recent tear of adding breaking changes for the 2.0.0 release, generally it's been when there's a compelling reason. As I'm not sure there's one here, I'm not convinced that replacing the old `Implements` is the way to go.

Sorry to try to squash your dreams. Perhaps there's an advantage I'm not seeing, other than the syntactic sugar? I'm always open to being convinced.

And, believe it or not, I (and the rest of the team, I'm sure) appreciate your input. It's great to see that you care about FakeItEasy.

**Comment by thomaslevesque** [comment]
> Unfortunately, I'm going to be my usual self and question the value.

Oh, I totally agree that the value is small ;). But the implementation cost is very small too. It's just that FIE usually has a really lean API, which I like, and having to type `Implements(typeof(IBar))` feels a bit wrong compared to the rest of the API.

> As I've recently been taking a fairly strong stance against introducing a second way of performing an operation when the new one lacks a sizable advantage over the original, I don't think I could advocate adding this mechanism as a second option for specifying Implements.

I get your point about not having several ways to do the same thing, but I'd argue that in that case it's not really a different way; you still call `Implements` to specify a type, you just do it with a slightly different syntax. Many libraries use both the generic and `typeof` approaches for specifying a type, and users are usually familiar with the two alternatives. It's always nice to have a choice.

> And as for replacing the original mechanism, while we've been on a recent tear of adding breaking changes for the 2.0.0 release, generally it's been when there's a compelling reason. As I'm not sure there's one here, I'm not convinced that replacing the old Implements is the way to go.

I wasn't thinking about replacing the current mechanism, just add a more convenient alternative to it; it wouldn't be a breaking change.

Do you mean that for 2.0 the API will still use `typeof` and only that ? I think that's a pity; if you're going to introduce breaking changes anyway, why not use the opportunity to make this part of the API better?

> Sorry to try to squash your dreams

Oh, it's not much of a dream ;). I just wanted to make a small inconvenience go away.

> Perhaps there's an advantage I'm not seeing, other than the syntactic sugar? I'm always open to being convinced.

Nope, just syntactic sugar.

Anyway, it's your project, so it's your call ;)

**Comment by blairconrad** [comment]
> I get your point about not having several ways to do the same thing, but I'd argue that in that case it's not really a different way; you still call Implements to specify a type, you just do it with a slightly different syntax. Many libraries use both the generic and typeof approaches for specifying a type, and users are usually familiar with the two alternatives.

I see your point. Let me consider that for a little bit.
However if we were to add the overload, I'd want to push the change into post-2.0.0 release. Since it's not breaking, we've no need to put into 2.0.0, and we already have a backlog of issues to be reviewed and merged for 2.0.0.

I'd really like to return to a "release every few weeks" schedule again once 2.0.0 is out…

**Comment by thomaslevesque** [comment]
Alright

**Comment by blairconrad** [comment]
Okay. I've taken it for a spin.

![image](https://cloud.githubusercontent.com/assets/3275797/6883307/68e390de-d57e-11e4-8867-fbfa04b925f5.png)

and

![image](https://cloud.githubusercontent.com/assets/3275797/6883316/d1c2c1ec-d57e-11e4-9265-abe7755ee9b4.png)

I don't think the addition of the new call will introduce confusion. As @thomaslevesque says, even though there are two ways to do it, this is not an uncommon pattern, and at least the two methods can be seen in the Intellisense together; it's not as if there are wildly divergent ways to do the same thing, for example as if the calls were accessed from different areas of the API.

If the other @FakeItEasy/owners have no objection, I say let's let the new overload in.

@thomaslevesque, I'm glad you took the time to reply to my knee-jerk rejection.

I'll make some comments in the PR, just in case the other owners also like the change. You may want to wait on implementing them until someone responds, though.

**Comment by blairconrad** [comment]
I'm content with the changes in the PR. If no owners dissent, I'm happy to merge it/have it merged at any time.

**Comment by adamralph** [comment]
I like the new overload. I have to agree that this has always been a slightly awkward item in the API. I've always known the type I wanted to additionally implement so having to use `typeof` always felt like an unnecessary loop to jump through.

I just have two questions a) just to clarify, is it right that we are talking about just adding the new overload here, not taking the old one away? I can see the current one may still have some uses if people need, for some reason or another, to use a local `Type` reference when making the call and b) how 'breaking' is the addition to the interface? Is that interface used anywhere in the extensibility points? Although I guess, since this will go out with 2.0, this doesn't really make any odds.

**Comment by thomaslevesque** [comment]
> a) just to clarify, is it right that we are talking about just adding the new overload here, not taking the old one away?

Yes, of course, the old one stays. There might be use cases where you can't use the generic form.

> b) how 'breaking' is the addition to the interface? Is that interface used anywhere in the extensibility points?

I thought about that. Adding a method to an existing public interface is usually something to be avoided, because it breaks existing implementations. However I can't think of any use case where someone would create its own implementation of `IFakeOptionsBuilder<T>`; the user just provides a delegate that uses an instance of it, he has no control over its creation. So in this case I think it's OK.

I would have preferred to make it an extension method, but it would need to have two type parameters, which would make it awkward to use...

**Comment by blairconrad** [comment]
I agree with @thomaslevesque, that this is very unlikely to break anything for anyone. I can't think why anyone would implement their own `IFakeOptionsBuilder<T>`.
Of course, if you've about 5 minutes to spare, @adamralph, or want to forgo the "four owner eyes" policy, I'm happy to merge right now and we'll have it in 2.0.0.

The implementation is (not to downplay @thomaslevesque's work) about as simple as can be imagined. And @thomaslevesque has graciously complied with all my requests regarding test coverage and even the CodeAnalysis/StyleCop warnings have been cleaned up.

**Comment by blairconrad** [comment]
@thomaslevesque, thanks very much for your work on this issue. Don't bother looking for your name in the release notes. :key: 

This issue has been fixed in [release 2.0.0](https://github.com/FakeItEasy/FakeItEasy/releases/tag/2.0.0).

**Comment by thomaslevesque** [comment]
> Don't bother looking for your name in the release notes. :key:

Oh :disappointed: 
:laughing:

**Comment by blairconrad** [comment]
@thomaslevesque, I was wondering if you were going to catch that. I did feel bad that you didn't make it onto the list, but I thought it would also look a little funny.
That's the price you pay for power, I guess.
